src/lib/message/L1ToL2Message.ts

```diff
@@ -228,18 +230,38 @@ export class L1ToL2MessageReader extends L1ToL2Message {
 
   // CHRIS: TODO: update these docs
   /**
-   * Receipt for the l2 transaction created by this message. See L1ToL2Message.l2TxHash
+   * Receipt for the first l2 transaction created by this message. See L1ToL2Message.l2TxHash
    * May throw an error if the l2 transaction has yet to be executed, which is the case if
    * the retryable ticket has not been created and redeemed.
-   * @returns
+   * @returns TransactionReceipt of the first redeem attempt if exists, otherwise null
    */
   public async getFirstRedeemAttempt(): Promise<TransactionReceipt | null> {
+    return await this.getFirstRedeem(undefined, true)
+  }
+
+    /**
+   * Receipt for the successful l2 transaction created by this message. See L1ToL2Message.l2TxHash
+   * May throw an error if the l2 transaction has yet to be executed, which is the case if
+   * the retryable ticket has not been created and redeemed.
+   * @returns TransactionReceipt of the first successful redeem if exists, otherwise null
+   */
+  public async getSuccessfulRedeemAttempt(): Promise<TransactionReceipt | null> {
```

Comment:
Maybe we can discuss later, but I think this is probably the only public function we need (although we should rename it to `getSuccessfulRedeem` since it only returns successful redeems). We may also want to keep `getAutoRedeem` if for some reason callers want to do a quick check of the auto redeem but are cautious about doing a heavy query, which the success case would entail.

Reply:
yea agreed that it can be useful to have a `getAutoRedeem`

Reply:
Agree that `getSuccessfulRedeemAttempt` should be the default, but for example retryable-dashboard would want `getFirstRedeemAttempt` in case there are no successful redeem so we can look at the revert reason. `getAutoRedeemAttempt` might be used to debug auto-redeem issue.

Reply:
Basically `getFirstRedeemAttempt`

Reply:
`getFirstRedeemAttempt` fallback to `getAutoRedeemAttempt` if there are auto-redeem attempt

Reply:
> But why is the first manual redeem attempt interesting, compared to for example the last one?

There are no guarantee `getSuccessfulRedeemAttempt` will return the last one, it may as well return nothing if there are no successful redeem attempt.

Say someone sent a L1 -> L2 tx that is both underpriced and may revert in L2. Since the auto-redeem fail, `getAutoRedeemAttempt` and `getSuccessfulRedeemAttempt` return null. If the user attempt to manual redeem, and the l2 tx reverted, he will need to use `getFirstRedeemAttempt` to find the reverting l2 tx.

Reply:
But if what you mean is to have `getSuccessfulRedeemAttempt` return the last redeem attempt (which the success attempt must be the last), we can do that too.
How about
`getSuccessfulRedeemAttempt` -> `getLastRedeem`
keep `getAutoRedeemAttempt` for debug 
drop `getFirstRedeemAttempt`  ?

Reply:
My concern is that `getLastRedeem` will force us to scan through all the block within `keepalive` for any failed retryable, but assuming the redeem is going to fail for similar reason, `getFirstRedeemAttempt` should be enough for one to debug what's wrong

Reply:
I'm ok with leaving `getSuccessfulRedeem` and just dropping the `getFirstRedeemAttempt`. We'll lose the ability to get the first manual redeem attempt (in the case that there wasnt an auto redeem), but my argument is that that's not really an interesting special case.

If we do want to provide info about the failing manual redeems, then I would suggest we focus on the last redeem, or provide access to all failed redeems. I think we can do that stuff in a later pr, if we choose to do it all.

Reply:
ok resolved in cad0fb7ac